pyquil/operator_estimation.py

```diff
@@ -799,17 +850,169 @@ def measure_observables(qc: QuantumComputer, tomo_experiment: TomographyExperime
+            continue
+        elif op_ch == 'X':
+            prog += Program(X(qubits[i]))
+        else:
+            raise ValueError("ops_strings_ should only consist of 'I's and/or 'X's")
+    return prog
+
+
+def _ops_str_to_flips(ops_str: str, qubits: List[int]) -> Dict:
+    """
+    :param ops_str: string specifying the operation to be carried out on `qubits`
+    :param qubits: list specifying the qubits to be carried operations on
+    :return: Dict specyfing whether to flip the readout results or not, depending on
+        the operations specified in `ops_str`, which in turn are operating on the
+        qubits specified in `qubits`
+    """
+    d_flip = {}
+    for i, op_ch in enumerate(ops_str):
+        q = qubits[i]
+        if op_ch == 'I':
+            d_flip[q] = 0
+        elif op_ch == 'X':
+            d_flip[q] = 1
+        else:
+            raise ValueError("ops_strings_ should only consist of 'I's and/or 'X's")
+    return d_flip
+
+
+def _stack_dicts(dict1: Dict, dict2: Dict) -> Dict:
+    """
+    :param dict1: Dict keyed with integer specifying qubit, valued by 1-dimensional numpy array specifying
+        readout results
+    :param dict2: Dict keyed with integer specifying qubit, valued by 1-dimensional numpy array specifying
+        readout results
+    :return: Dict keyed with integer specifying qubit, valued by 1-dimensional numpy array specifying
+        readout results gathered from both `dict1` and `dict2`
+    """
+    assert set(dict1.keys()) == set(dict2.keys()), "Dictionaries must have same keys"
+    assert set(len(v.shape) for v in dict1.values()) == set(len(v.shape) for v in dict2.values()), \
+        "Arrays in dict values must have same dimension"
+    dict_combined = {}
+    for k, v in dict1.items():
+        dict_combined[k] = np.hstack([v, dict2[k]])
+    return dict_combined
+
+
+def _stats_from_measurements(d_results: dict, setting: ExperimentSetting,
+                             n_shots: int, coeff: float = 1.0) -> Tuple[float]:
+    """
+    :param d_results: results from running `qc.run_and_measure()`
+    :param setting: ExperimentSetting
+    :param n_shots: number of shots in the measurement process
+    :param coeff: coefficient of the operator being estimated
+    :return: tuple specifying (mean, variance)
+    """
+    # Transform bits to eigenvalues; ie (+1, -1)
+    obs_strings = {q: 1 - 2 * d_results[q] for q in d_results}
+    # Pick columns corresponding to qubits with a non-identity out_operation and stack
+    # into an array of shape (n_shots, n_measure_qubits)
+    my_obs_strings = np.vstack([obs_strings[q] for q, op_str in setting.out_operator]).T
+    # Multiply row-wise to get operator values. Do statistics. Return result.
+    obs_vals = coeff * np.prod(my_obs_strings, axis=1)
+    obs_mean = np.mean(obs_vals)
+    obs_var = np.var(obs_vals) / n_shots
+
+    return obs_mean, obs_var
+
+
+def ratio_variance(a: float, var_a: float, b: float, var_b: float) -> float:
```

Comment:
private function?

Reply:
Maybe...I can see this function being used on its own elsewhere too, so I'm inclined to keep it non-private. What wasn't very obvious before is that it can also work with numpy arrays and not just floats, so I've changed the input types, and added a few tests to make this more obvious.